### DISCUSSION TITLE
It is possible to mock file modules?

### DISCUSSION THREAD

**Discussion by 128f** [asomers/mockall#358] [Questions] [answered]
I have a file `somefile` that is just a set of functions.
I am using `pub mod somefile`, and then doing `use somefile::{foo, bar}`.
I want to mock the functions coming from the file, but I don't want to rewrite the interface.

I have seen https://github.com/asomers/mockall/issues/219, but it seems the suggestion there is to rewrite all function signatures in the same place I do `pub mod somefile`.

My only option seems to be to introduce an internal module `baz` into `somefile`,
which I can then apply `cfg_attr(test, automock)` to,
and then change my references to `use somefile::baz::{foo,bar}`.

Is this the only way? Or is there a way to automock `somefile` without rewriting the interface to it?

**Comment by asomers** [comment, marked as answer]
You could do something like this in somefile.rs:
```rust
// somefile.rs
use cfg_if::cfg_if;
use mockall::automock;

#[automock]
pub mod inner {
    pub fn foo(x: u32) -> u32 {...}
}

cfg_if! {
    if #[cfg(test)] {
        pub use mock_inner::foo;
    } else {
        pub use inner::foo;
    }
}
```
That approach has the advantage that you don't have to modify any other file.  Alternatively, if you really don't want to modify somefile.rs, you could create mock_somefile.rs which repeats the function definitions like this, and then use `#[double]` in the other files


```rust
// mock_somefile.rs
use mockall::automock;

#[automock]
pub mod inner {
    pub fn foo(x: u32) -> u32 {unimplemented!()}
}
```

**Reply by 128f** [reply]
Your first thing was almost exactly what I was looking for...
I could probably do something like `pub use mock_inner::*;`, right?

But what about modules from external crates, where I can't make edits introducing an intermediate module?
Is that OOS/unimplemented in this crate?

**Reply by asomers** [reply]
Yes, glob imports will work.  As for modules from external crates, you must repeat the definition somehow.  Probably something close to my second example is what you want.
